enterprise/cli/dbcrypt_rotate.go

```diff
@@ -0,0 +1,152 @@
+				if err != nil {
+					return xerrors.Errorf("key must be base64-encoded")
+				}
+				if idx == 0 {
+					newKey = dk
+				} else if bytes.Equal(dk, newKey) {
+					return xerrors.Errorf("old key at index %d is the same as the new key", idx)
+				}
+				keys = append(keys, dk)
+			}
+
+			ciphers, err := dbcrypt.NewCiphers(keys...)
+			if err != nil {
+				return xerrors.Errorf("create ciphers: %w", err)
+			}
+
+			sqlDB, err := cli.ConnectToPostgres(inv.Context(), logger, "postgres", vals.PostgresURL.Value())
+			if err != nil {
+				return xerrors.Errorf("connect to postgres: %w", err)
+			}
+			defer func() {
+				_ = sqlDB.Close()
+			}()
+			logger.Info(ctx, "connected to postgres")
+
+			db := database.New(sqlDB)
+
+			cryptDB, err := dbcrypt.New(ctx, db, ciphers...)
+			if err != nil {
+				return xerrors.Errorf("create cryptdb: %w", err)
+			}
+
+			users, err := cryptDB.GetUsers(ctx, database.GetUsersParams{})
+			if err != nil {
+				return xerrors.Errorf("get users: %w", err)
+			}
+			logger.Info(ctx, "encrypting user tokens", slog.F("count", len(users)))
+			for idx, usr := range users {
+				err := cryptDB.InTx(func(tx database.Store) error {
+					userLinks, err := tx.GetUserLinksByUserID(ctx, usr.ID)
+					if err != nil {
+						return xerrors.Errorf("get user links for user: %w", err)
+					}
+					for _, userLink := range userLinks {
+						if _, err := tx.UpdateUserLink(ctx, database.UpdateUserLinkParams{
+							OAuthAccessToken:  userLink.OAuthAccessToken,
+							OAuthRefreshToken: userLink.OAuthRefreshToken,
+							OAuthExpiry:       userLink.OAuthExpiry,
+							UserID:            usr.ID,
+							LoginType:         usr.LoginType,
+						}); err != nil {
+							return xerrors.Errorf("update user link: %w", err)
+						}
+					}
+					gitAuthLinks, err := tx.GetGitAuthLinksByUserID(ctx, usr.ID)
+					if err != nil {
+						return xerrors.Errorf("get git auth links for user: %w", err)
+					}
+					for _, gitAuthLink := range gitAuthLinks {
+						if _, err := tx.UpdateGitAuthLink(ctx, database.UpdateGitAuthLinkParams{
+							ProviderID:        gitAuthLink.ProviderID,
+							UserID:            usr.ID,
+							UpdatedAt:         gitAuthLink.UpdatedAt,
+							OAuthAccessToken:  gitAuthLink.OAuthAccessToken,
+							OAuthRefreshToken: gitAuthLink.OAuthRefreshToken,
+							OAuthExpiry:       gitAuthLink.OAuthExpiry,
+						}); err != nil {
+							return xerrors.Errorf("update git auth link: %w", err)
+						}
+					}
+					return nil
+				}, &sql.TxOptions{
+					Isolation: sql.LevelRepeatableRead,
+				})
+				if err != nil {
+					return xerrors.Errorf("update user links: %w", err)
+				}
+				logger.Debug(ctx, "encrypted user tokens", slog.F("current", idx+1), slog.F("cipher", ciphers[0].HexDigest()))
+			}
+			logger.Info(ctx, "operation completed successfully")
+
+			// Revoke old keys
+			for _, c := range ciphers[1:] {
+				if err := db.RevokeDBCryptKey(ctx, c.HexDigest()); err != nil {
```

Comment:
This could fail if a coder server was still running with the old key as the primary key right? Is there any way to defend against that?

Reply:
Only if it writes some new encrypted data with the old key between the time that we re-encrypted and the time we attempt to revoke. And in this case, it would be a legit error - operators should be updating the deployment with the new keys before rotating or stopping deployments without access to the new keys.

Reply:
Yeah I understand that, but is there any way we can defend against trying to rotate while the key is still in use by a running coder replica?

Reply:
A heartbeat column on the keys table comes to mind, and each replica can update it's primary key every 5 minutes or so. And then require waiting 10 minutes since last seen before you can rotate it out of service

Reply:
I'm not worried about this command _failing_ --- the bigger problem is if you screw up the keys, this command _succeeds_, and you've encrypted the database with a key your coderds don't have.  That's presumably an outage.

That's one advantage of moving this to an API call --- the Coderd serving the API can use they key it already has.  There is still a risk of locking out other coderds, but at least you know one coderd can decrypt.

Reply:
You could also use the heartbeat system to ensure that at least one replica has the key too.

I think doing it as an API call adds a lot of complexity and causes more issues. Large databases will time out the request, so it needs to be implemented as a background task. Background tasks in coderd don't really have a way to notify on success/failure without a lot of new plumbing, and if the replica gets stopped for any reason any running goroutines get canceled after a short period of time to stop blocking the shutdown and avoid a sigkill.

I think doing it as a subcommand avoids all of these complexities. You can still avoid the missing key problem with a bit of thought into the keys table so I don't think this is a good reason to do it as an API call.

Reply:
The API could stream progress responses to avoid the client thinking it has hung and timing out the request.  Background task is not necessary. Working in transactions prevents any partially-complete work from corrupting the database, so replica getting stopped isn't a big deal either.  You just re-run the rotation on a different replica.

Reply:
If a key is already in use by another coder replica, and we attempt to rotate the key, one of two things will happen:
1. The rotation fails to complete, in which case the key will remain in rotation.
2. The rotation will complete successfully, in which case the key will be revoked.

In the case of 1), the administrator still has the option to stop the old instance and retry rotation.
In the case of 2), any attempts by the old instance to insert new data with the old key will fail due to the foreign key constraint.

I don't think we need to add heartbeats, but I specifically call out in the documentation to add the new key to running instances _before_ performing rotation.